contracts/poap/src/contract.rs

```diff
@@ -0,0 +1,1375 @@
+                current_time: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS),
+            },
+            result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn event_info_start_time_equal_current_time_error() {
+        let mut deps = mock_dependencies();
+        let mut env = mock_env();
+        let info = mock_info(CREATOR, &vec![]);
+
+        do_instantiate(deps.as_mut());
+        env.block.time = Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS);
+
+        let msg = ExecuteMsg::UpdateEventInfo {
+            start_time: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS),
+            end_time: Timestamp::from_seconds(EVENT_END_SECONDS),
+        };
+
+        let result = execute(deps.as_mut(), env.clone(), info.clone(), msg);
+        assert_eq!(
+            ContractError::StartTimeBeforeCurrentTime {
+                start_time: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS),
+                current_time: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS),
+            },
+            result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn event_info_end_time_before_current_time_error() {
+        let mut deps = mock_dependencies();
+        let mut env = mock_env();
+        let info = mock_info(CREATOR, &vec![]);
+
+        do_instantiate(deps.as_mut());
+        env.block.time = Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS);
+
+        let msg = ExecuteMsg::UpdateEventInfo {
+            start_time: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS + 2),
+            end_time: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS - 1),
+        };
+
+        let result = execute(deps.as_mut(), env.clone(), info.clone(), msg);
+        assert_eq!(
+            ContractError::StartTimeAfterEndTime {
+                start: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS + 2),
+                end: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS - 1),
+            },
+            result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn event_info_end_time_equal_current_time_error() {
+        let mut deps = mock_dependencies();
+        let mut env = mock_env();
+        let info = mock_info(CREATOR, &vec![]);
+
+        do_instantiate(deps.as_mut());
+        env.block.time = Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS);
+
+        let msg = ExecuteMsg::UpdateEventInfo {
+            start_time: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS + 2),
+            end_time: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS),
+        };
+
+        let result = execute(deps.as_mut(), env.clone(), info.clone(), msg);
+        assert_eq!(
+            ContractError::StartTimeAfterEndTime {
+                start: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS + 2),
+                end: Timestamp::from_seconds(INITIAL_BLOCK_TIME_SECONDS),
+            },
+            result.unwrap_err()
+        );
+    }
+
+    #[test]
+    fn update_admin_only_from_admin() {
```

Comment:
I think it can be splited to:
1. update_admin_without_permission_error
2. update_admin_with_permission_properly